### ISSUE TITLE
Cannot learn initial_states with batch_sequences_with_states

### ISSUE THREAD

**Issue by rubenvereecken** [tensorflow/tensorflow#12086]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Mint 18
- **TensorFlow installed from (source or binary)**: Binary (pip)
- **TensorFlow version (use command below)**: v1.3.0.0rc0
- **Bazel version (if compiling from source)**: N/A
- **CUDA/cuDNN version**: N/A
- **GPU model and memory**: N/A
- **Exact command to reproduce**: N/A

### Describe the problem
Using `tf.contrib.training.batch_sequences_with_states`, it seems impossible to learn the `initial_states` passed to it. I've tried using Variables with `trainable=True` with several initializers and while my optimizer picks them up through `tf.trainable_variables()`, their values do not get learned.

I assume this has something to do with TensorFlow continuing to work with the Tensors, not the Variables? Additionally, the initializer is called each run and in TensorBoard it shows that it does not feed into the optimizer, whereas network weights for example would.

Are my presumptions correct, and if not, how do I fix this? If this is a TensorFlow limitation, how could I work around it?

### Source code / logs
```
            initializers = {
                    'lstm_c': train_init,
                    'lstm_h': train_init,
                    'encode_lstm_c': train_init,
                    'encode_lstm_h': train_init,
                    'last_out': train_init,
                    }

            with tf.variable_scope(tf.get_variable_scope(), reuse=not is_training):

                initial_states = { k: tf.get_variable('initial_{}'.format(k), v.shape,
                        dtype=v.dtype,
                        initializer=initializers[k],
                        trainable=is_training and self.learn_initial_states, \
                        collections=[attend.GraphKeys.INITIAL_STATES]) \
                        for k, v in initial_constants.items() }


                # This makes sure we're not dealing with the reference but learned values
                initial_states = { k: v.initialized_value() for k, v in initial_states.items() }
```

**Comment by rohan100jain** [comment]
I'm not sure I completely understand your question... The documentation for NextQueuedSequenceBatch (object returned from tf.contrib.training.batch_sequences_with_states) seems to suggest that the state() and save_state() methods could be used to see / store the values of the states. 

Also, I think this question might be more suitable for StackOverflow.

**Comment by rubenvereecken** [comment]
I agree I worded it as a SO question but that's only because of my frustration to get this to work and an inability to understand the inner workings of variables in this scenario. In all other places I would just create a variable and multiple it for example and the optimizer would know what to do. In this scenario it's a lot less clear and I can't use variables inside `initial_states` passed to `batch_sequences_with_state`,  I have to get their `initialized_value`. 

I would also like to discuss this use case if it turns out it's currently not supported and request support for it, or discuss the workaround that should commonly be adopted. Learning initial values for for example recurrent layers is a fairly common practice I've gathered.

**Comment by milah** [comment]
You are right the initial_states cannot be learned with the SequenceQueueingStateSaver that is constructed in batch_sequences_with_states.

No gradients can flow through the queue. This is a limitation. The documentation could be a bit clearer but it mentions "e.g. constants or tensors" for the initial state.

The only way around this is not using the SequenceQueueingStateSaver. For instance, how about training the initial state with the dynamic_rnn on small enough (pre-segmented) sequences? Then you can switch to the state_saving_rnn with the SequenceQueueingStateSaver with the pre-trained initial state.

**Comment by rubenvereecken** [comment]
I have also tried something with `tf.cond` and `tf.map_fn` to conditionally use variables instead of the tensor states provided by the state saver when it notices a certain batch is the first one in its sequence, but to no avail.

Pre-training indeed sounds like the only way to go about this. Not a bad idea at all, thanks. I'd still hoped to make this work end-to-end.

**Comment by rubenvereecken** [comment]
Never mind, my last attempt didn't work because I messed up variable initialization.

The way I worked around this is as follows. I keep a `first` boolean in my state saver which is initialized to `True` and gets set to `False` after the first time a batch is encountered.
I keep a separate set of to-be-learned initial states. The trick is to only use them during the first batch for a sequence. I do this by creating a mask from `first`, expanding the initial state vector to the correct batch size, masking it, then adding it to the state saver's initial states, effectively replacing all initial zero rows and only those. This will include the initial state variables in the computation graph making them optimizable.

```
             c = state_saver.state('lstm_c')
             first = state_saver.state('first')

             only_first_mask = tf.expand_dims(first, 1) # Mask rows
             not_first_mask = tf.logical_not(only_first_mask)
             only_first_mask = tf.cast(only_first_mask, tf.float32)
             not_first_mask = tf.cast(not_first_mask, tf.float32)

             # learnable initial variable
             init_c = provider.initial_variables['lstm_c']
             # B x 1 * 1 x 512 -> 512 x 512 (repeats row)
             # Repeats the initial variable `batch_size` times
             init_c = tf.einsum('ij,jk->ik', tf.ones([batch_size,1]), tf.expand_dims(init_c,0))
             # Mask the initial variable
             init_c = only_first_mask * init_c
             c = not_first_mask * c # Has first=true set to zeros
             # This now contains entries from `init_c` in the rows corresponding to `first`
             c = c + init_c
```
